pubtools/pulplib/_impl/client/client.py

```diff
@@ -179,6 +182,28 @@ def search_repository(self, criteria=None):
             response_f, lambda data: self._handle_page(Repository, search, data)
         )
 
+    def _do_upload_file(self, upload_id, file_obj):
+        def do_next_upload(checksum, size):
+            data = file_obj.read(self._CHUNK_SIZE)
```

Comment:
Is there a problem here with how file_obj is managed?  Here, you assume file_obj is already a file object, but you may coerce it to a file object *after* do_next_upload is declared.  Does that actually work?  i.e. does this nested function capture file_obj at the point where do_next_upload is declared, in which case it's wrong, or does it somehow hold a reference to the outer scope which means it can see subsequent assignments to file_obj?

This isn't a rhetorical question as I'm not familiar with all details how closures work in Python.  If you're not sure, please test it.  Or even just refactor the code to avoid this - I often find a little callable helper class (replacing `do_next_upload` with a `class Upload` having a `__call__` method) ends up with more readable code.

Reply:
I think this is Okay, the function defined inside a function can access variables defined outside its body and it treats these variables as 'free variables', hold the references to them.
Here's a simple example, shows it keeps the reference to the variable. 
```
>>> def f():
...     def f1():
...             print(a)
...     a = 1
...     print(hex(id(a)))
...     return f1
... 
>>> p = f()
0x7f0c433974e0
>>> p.__closure__
(<cell at 0x7f0c35c93708: int object at 0x7f0c433974e0>,)
```

Reply:
Thanks Jay.  Repeated it locally on both python2 and python3, looks like it works fine.